Add documentation for EIB's RPM resolution logic (#161)
suse-edge/edge-image-builder: docs/design/pkg-resolution.md

# Package resolution design
The package resolution design can be separated in three logical parts:
1. *Running the EIB container* - how to run the EIB container so that the **RPM resolution** has the needed permissions
1. *Building the EIB image* - what happens during the **RPM resolution** logic of EIB's image build
1. *Booting the EIB image* - how are the packages actually installed once the EIB image is booted for the first time

## Running the EIB container
![image](../images/rpm-eib-container-run.png)

> The package resolution workflow begins with the user specifying packages and/or stand-alone RPMs that will be installed when the EIB image is booted. On how to to do a correct specification, see [Specify packages for installation](installing-packages.md#specify-packages-for-installation).

Reviewer: The URL is now broken.

> After the desired specification has been made, the user runs the EIB container with the [`--privileged`](https://docs.podman.io/en/latest/markdown/podman-run.1.html#privileged) option, ensuring that EIB has the needed permissions to successfully run a Podman instance within its container. This is a crutial prerequisite for building a working EIB image with package installation specified (more on this in the next section). 

Reviewer: ```suggestion
After the desired specification has been made, the user runs the EIB container with the [`--privileged`](https://docs.podman.io/en/latest/markdown/podman-run.1.html#privileged) option, ensuring that EIB has the needed permissions to successfully run a Podman instance within its container. This is a crucial prerequisite for building a working EIB image with package installation specified (more on this in the next section). 
```

podman run --rm --privileged -it \
-v $IMAGE_DIR:/eib eib:dev /bin/eib build \
--config-file $CONFIG_FILE.yaml \
--config-dir /eib \
--build-dir /eib/_build
```

> **_NOTE:_** Depending on the `cgroupVersion` that Podman operates with, you might also need to run the command with `root` permissions. This is the case for `cgroupVersion: v1`. In this version, non-root usage of the `--privileged` option is not supported. For `cgroupVersion: v2`, non-root usage is supported. 
>
>In order to check the `cgroupVersion` that Podman operates with, run the following command:
>```shell
>podman info | grep cgroupVersion
>```

Once the EIB container has been successfully executed, it parses all the user provided specification and begins the **RPM resolution** process. 

## Building the EIB image
During this phase, EIB prepares the user specified packages for installation. This process is called **RPM resolution** and it includes:
1. Validating that each provided package has a GPG signature or comes from a GPG signed RPM repository
1. Resolving and downloading the dependencies for each specified package
1. Creating a RPM repository that consists of the specified packages and their dependencies

> 1. Configure the usage of this repositry for package installation during the **combustion** phase of the EIB image boot

Reviewer: Consistency / styling suggestion:

```suggestion
1. Configuring the usage of this repositry for package installation during the **combustion** phase of the EIB image boot
```

### RPM resolution process
![image](../images/rpm-resolver-architecture.png)

> EIB mainly utilizes Podman's functionality to setup the environment needed for the **RPM resolution** process. In order to communicate with Podman, EIB first creates a [listening service](https://docs.podman.io/en/latest/markdown/podman-system-service.1.html) that will faciliate the communication between EIB and Podman. From here onwards, asume that any Podman related operation that EIB does goes through the **listening service** first.

Reviewer: ```suggestion
EIB mainly utilizes Podman's functionality to setup the environment needed for the **RPM resolution** process. In order to communicate with Podman, EIB first creates a [listening service](https://docs.podman.io/en/latest/markdown/podman-system-service.1.html) that will facilitate the communication between its own container and Podman. From here onwards, assume that any Podman related operation that EIB does goes through the **listening service** first.
```

The final step in the EIB **RPM resolution** process is to create an **install script** which uses the aforementioned RPM repository to install the user specified packages during the EIB image combustion phase.

### Troubleshooting
When troubleshooting the **RPM resolution** process, it is beneficial to look at the following files/directories inside of the EIB build directory:
1. `eib-build.log` - general logs for the whole EIB image build process
1. `podman-image-build.log` - logs for the build of the EIB resolver image. If missing, but the `resolver-image-build` directory is present, this means that there is a problem in the configuration of the `resolver-image-build` directory
1. `podman-system-service.log` - logs for the Podman listening service
1. `resolver-image-build` directory - build context for the resolver image. Make sure that the `Dockerfile` holds correct data. When installing side-loaded RPMs, make sure that the `rpms` and `gpg-keys` directories are present in the `resolver-image-build` directory

> 1. `resolver-tarball-image` direcotry - contains resources related to the creation of the **virtual disk tarball** archive. If this directory exists, this means that a problem has been encountered while EIB was trying to import the **tarball image**

Reviewer: ```suggestion
1. `resolver-tarball-image` directory - contains resources related to the creation of the **virtual disk tarball** archive. If this directory exists, this means that a problem has been encountered while EIB was trying to import the **tarball image**
```
